drivers/st25r3911b/st25r3911b.c

```diff
@@ -0,0 +1,570 @@
+
+	return 0;
+}
+
+static int measure_voltage(u8_t source, u32_t *voltage)
+{
+	int err;
+	u8_t reg;
+	u32_t result;
+	u32_t status;
+
+	source &= ST25R3911B_REG_REGULATOR_CTRL_MPSV_MASK;
+
+	/* Set measure source to VDD. */
+	err = st25r3911b_reg_modify(ST25R3911B_REG_REGULATOR_CTRL,
+				    ST25R3911B_REG_REGULATOR_CTRL_MPSV_MASK,
+				    source);
+	if (err) {
+		return err;
+	}
+
+	err = st25r3911b_irq_enable(ST25R3911B_IRQ_MASK_DCT);
+	if (err) {
+		return err;
+	}
+
+	LOG_DBG("Power supply voltage measurement");
+
+	err = st25r3911b_cmd_execute(ST25R3911B_CMD_MEASURE_POWER_SUPPLY);
+	if (err) {
+		st25r3911b_irq_disable(ST25R3911B_IRQ_MASK_DCT);
+		return err;
+	}
+
+	status = st25r3911b_irq_wait_for_irq(ST25R3911B_IRQ_MASK_DCT,
+					     T_POWER_SUPP_MEAS);
+
+	err = st25r3911b_irq_disable(ST25R3911B_IRQ_MASK_DCT);
+	if (err) {
+		return err;
+	}
+
+	if (!status) {
+		return -ENXIO;
+	}
+
+	/* Get result */
+	err = st25r3911b_reg_read(ST25R3911B_REG_AD_CONVERTER_OUT, &reg);
+	if (err) {
+		return err;
+	}
+
+	/* Calculate voltage level in mV */
+	result = reg * POWER_SUPP_MEAS_INTEGER;
+	result += ((reg * POWER_SUPP_MEAS_FRACTION) +
+		    POWER_SUPP_MEAS_FRACTION_DIVISOR / 2) /
+		    POWER_SUPP_MEAS_FRACTION_DIVISOR;
+
+	*voltage = result;
+
+	LOG_DBG("Measured supply voltage %u mV", result);
+
+	return 0;
+}
+
+static int calibrate_antenna(void)
+{
+	int err;
+	u8_t reg = 0;
+	u32_t status;
+
+	err = st25r3911b_reg_read(ST25R3911B_REG_ANTENNA_CAL_CTRL, &reg);
+	if (err) {
+		return err;
+	}
+
+	/* Check if is possible to automatic calibration. */
+	if (reg & ST25R3911B_REG_ANTENNA_CAL_CTRL_TRIM_S) {
+		LOG_ERR("Antenna auto calibration is not possible");
+		return -EACCES;
+	}
+
+	err = st25r3911b_irq_enable(ST25R3911B_IRQ_MASK_DCT);
+	if (err) {
+		return err;
+	}
+
+	err = st25r3911b_cmd_execute(ST25R3911B_CMD_CALIBRATE_ANTENNA);
+	if (err) {
+		st25r3911b_irq_disable(ST25R3911B_IRQ_MASK_DCT);
+		return err;
+	}
+
+	LOG_DBG("Wait for antenna calibration %u ms", T_COMMON_CMD);
+
+	status = st25r3911b_irq_wait_for_irq(ST25R3911B_IRQ_MASK_DCT,
+					     T_COMMON_CMD);
+
+	err = st25r3911b_irq_disable(ST25R3911B_IRQ_MASK_DCT);
+	if (err) {
+		return err;
+	}
+
+	if (!status) {
+		return -ENXIO;
+	}
+
+	LOG_DBG("Antenna calibration success");
+
+	return 0;
+}
+
+static int rx_tx_enable(void)
+{
+	u8_t mask = ST25R3911B_REG_OP_CTRL_TX_EN |
+		    ST25R3911B_REG_OP_CTRL_RX_EN;
+
+	return st25r3911b_reg_modify(ST25R3911B_REG_OP_CTRL, 0, mask);
+}
+
+static int adjust_regulator(u32_t *mV_adjust)
+{
+	int err;
+	u8_t reg;
+	u8_t voltage = 0;
+	u8_t conf = 0;
+	u32_t mV;
+	u32_t status;
+
+	err = st25r3911b_reg_read(ST25R3911B_REG_REGULATOR_CTRL, &reg);
+	if (err) {
+		return err;
+	}
+
+	/* Check possibility of auto calibration. This bit should be 0 */
+	if (reg & ST25R3911B_REG_REGULATOR_CTRL_REG_S) {
+		LOG_ERR("Regulator auto calibration is not possible");
+		return -EACCES;
+	}
+
+	err = st25r3911b_irq_enable(ST25R3911B_IRQ_MASK_DCT);
```

Comment:
the are several functions that execute the following sequence:
```
st25r3911b_irq_enable(x);
if (st25r3911b_cmd_execute(y))
   st25r3911b_irq_disable(x);
status = st25r3911b_irq_wait_for_irq(x, z);
st25r3911b_irq_disable(x);
```

maybe this could be extracted to a common function?

Reply:
Ok